src/ir/can_derive_debug.rs

```diff
@@ -0,0 +1,274 @@
+
+        match *ty.kind() {
+            TypeKind::Void |
+            TypeKind::NullPtr |
+            TypeKind::Int(..) |
+            TypeKind::Float(..) |
+            TypeKind::Complex(..) |
+            TypeKind::Function(..) |
+            TypeKind::Enum(..) |
+            TypeKind::Reference(..) |
+            TypeKind::BlockPointer |
+            TypeKind::Named |
+            TypeKind::UnresolvedTypeRef(..) |
+            TypeKind::ObjCInterface(..) |
+            TypeKind::ObjCId |
+            TypeKind::ObjCSel => {
+                false
+            },
+            TypeKind::Opaque => {
+                if ty.layout(self.ctx)
+                    .map_or(true, |l| l.opaque().can_derive_debug(self.ctx, ())) {
+                        return false;
+                    } else {
+                        self.cant_derive_debug.insert(id);
+                        return true;
+                    }
+            },
+            TypeKind::Array(t, len) => {
+                if len <= RUST_DERIVE_IN_ARRAY_LIMIT {
+                    if self.cant_derive_debug.contains(&t) {
+                        self.cant_derive_debug.insert(id);
+                        return true;
+                    }
+                    false
+                } else {
+                    self.cant_derive_debug.insert(id);
+                    true
+                }
+            },
+            TypeKind::ResolvedTypeRef(t) |
+            TypeKind::TemplateAlias(t, _) |
+            TypeKind::Alias(t) => {
+                if self.cant_derive_debug.contains(&t) {
+                    self.cant_derive_debug.insert(id);
+                    return true
+                }
+                false
+            },
+            TypeKind::Comp(ref info) => {
+                if info.has_non_type_template_params() {
+                    if ty.layout(self.ctx).map_or(false,
+                                                           |l| l.opaque().can_derive_debug(self.ctx, ())) {
+                        return false;
+                    } else {
+                        self.cant_derive_debug.insert(id);
+                        return true;
+                    }
+                }
+                if info.kind() == CompKind::Union {
+                    if self.ctx.options().unstable_rust {
+                        self.cant_derive_debug.insert(id);
+                        return true;
+                    }
+
+                    if ty.layout(self.ctx).unwrap_or_else(Layout::zero)
+                        .opaque()
+                        .can_derive_debug(self.ctx, ()) {
+                            return false;
+                        } else {
+                            self.cant_derive_debug.insert(id);
+                            return true;
+                        }
+                }
+                let consider_base = info.base_members()
+                    .iter()
+                    .all(|base| !self.cant_derive_debug.contains(&base.ty));
```

Comment:
Each of the places where we do something like

```rust
let things_can_derive = things.all(|thing| !self.cant_derive_debug.contains(thing));
```

should become

```rust
let things_cant_derive = things.any(|thing| self.cant_derive_debug.contains(thing));
```

Its equivalent (de Morgan's) but the latter reads much more naturally.

Additionally, we should early return whenever we find something that would cause us to update the `cant_derive_debug` set, rather than continuing to consider (for example) fields even after we know we can't derive debug because of the bases.

That is, we should do something like this:

```rust
let bases_cant_derive = ...;
if bases_cant_derive {
    self.cant_derive_debug.insert(id);
    return true;
}

let fields_cant_derive = ...;
if fields_cant_derive {
    self.cant_derive_debug.insert(id);
    return true;
}

false
```

To avoid repeating the insert-into-the-set-and-then-return-true dance over and over, we can define a helper:

```rust
impl CantDeriveDebugAnalysis {
    ...
    fn insert(&mut self, id: ItemId) -> bool {
        let was_already_in = self.cant_derive_debug.insert(&id);
        assert!(
            !was_already_in,
            "We shouldn't try and insert twice because if it was already in the set, \
             `constrain` would have exited early."
        );
        true
    }
}
```